Mathlib/Topology/Covering/Lift.lean

```diff
@@ -0,0 +1,249 @@
+          exact Φ.curry.continuous
+    · rintro ⟨⟨γ, e⟩, hγ, he⟩
+      dsimp
+      constructor
+      · rw [concatCM_left (S.ht n.zero_le)]
+        exact hΦ ⟨⟨γ, e⟩, hγ, he⟩ |>.1
+      · rintro ⟨t, ht⟩
+        by_cases htn : t ≤ S.t n
+        · rw [concatCM_left htn] ; exact hΦ ⟨⟨γ, e⟩, hγ, he⟩ |>.2 ⟨t, _⟩
+        · rw [concatCM_right <| le_of_not_le htn]
+          simp only [comp_apply, coe_mk, Trivialization.proj_clift (proj := f)]
+          rfl
+
+/-- Lifting paths through a `LiftSetup`, as a bundled continuous map from paths adapted to the
+setup. -/
+noncomputable def lift (S : LiftSetup f) :
+    {F : C(S.Liftable, C(I, E)) // ∀ γe, F γe 0 = γe.1.2 ∧ ∀ t, f (F γe t) = γe.1.1 t} := by
+  obtain ⟨F, hF⟩ := S.partial_lift S.n le_rfl
+  let Φ : C(I, Icc (S.t 0) (S.t S.n)) := ⟨fun t => ⟨t, by simp⟩, by fun_prop⟩
+  refine ⟨⟨fun γe => (F γe).comp Φ, by fun_prop⟩, fun γe => ⟨?_, fun t => ?_⟩⟩
+  · simpa [Bot.bot] using hF γe |>.1
+  · simpa [embed] using hF γe |>.2 (Φ t)
+
+end LiftSetup
+
+include hf
+
+theorem exists_unique_lift (he : f e = γ 0) : ∃! Γ : C(I, E), Γ 0 = e ∧ f ∘ Γ = γ := by
+  obtain ⟨S, hS⟩ := LiftSetup.exist hf γ
+  obtain ⟨F, hF⟩ := S.lift
+  have h1 : F ⟨⟨γ, e⟩, hS, he⟩ 0 = e := hF ⟨⟨γ, e⟩, hS, he⟩ |>.1
+  have h2 : f ∘ F ⟨⟨γ, e⟩, hS, he⟩ = γ := by ext t ; exact hF ⟨⟨γ, e⟩, hS, he⟩ |>.2 t
+  refine ⟨F ⟨⟨γ, e⟩, hS, he⟩, ⟨h1, h2⟩, ?_⟩
+  rintro Γ ⟨rfl, hΓ₂⟩
+  apply coe_injective
+  apply hf.eq_of_comp_eq (a := 0) Γ.continuous (ContinuousMap.continuous _) <;> simp [*]
+
+/-- The path obtained by lifting through a covering map. -/
+noncomputable def lift (γ : C(I, X)) (he : f e = γ 0) : C(I, E) :=
+  (hf.exists_unique_lift he).choose
+
+@[simp]
+theorem lift_spec (γ : C(I, X)) (he : f e = γ 0) : hf.lift γ he 0 = e ∧ f ∘ hf.lift γ he = γ :=
+  (hf.exists_unique_lift he).choose_spec.1
+
+variable {Y : Type*} [TopologicalSpace Y] {γ : C(I × Y, X)} {Γ₀ : C(Y, E)}
+
+private def slice (γ : C(I × Y, X)) : C(Y, C(I, X)) := γ.comp prodSwap |>.curry
+
+private noncomputable def joint_lift (hΓ₀ : ∀ y, f (Γ₀ y) = γ (0, y)) : C(Y, C(I, E)) := by
+  use fun y => hf.lift (slice γ y) (hΓ₀ y)
+  rw [continuous_iff_continuousAt]
+  intro y₀
+  obtain ⟨S, hS⟩ := LiftSetup.exist hf (slice γ y₀)
+  apply ContinuousOn.continuousAt ?_ hS.eventually
+  rw [continuousOn_iff_continuous_restrict]
+  obtain ⟨G₁, hG₁⟩ := S.lift
+  let G₂ : C({y // S.fits (slice γ y)}, S.Liftable) :=
+    ⟨fun y => ⟨⟨slice γ y, Γ₀ y⟩, y.2, hΓ₀ y⟩, by fun_prop⟩
+  convert G₁.comp G₂ |>.continuous
+  ext1 y
+  have h3 := hG₁ ⟨⟨slice γ y, Γ₀ y⟩, y.2, hΓ₀ y⟩
+  apply coe_injective
+  apply hf.eq_of_comp_eq (a := 0) (ContinuousMap.continuous _) (ContinuousMap.continuous _)
+  · simp [G₂, h3]
+  · simp only [Set.restrict_apply, mem_setOf_eq, hf.lift_spec, comp_apply, coe_mk, G₂]
+    ext t
+    simp [h3]
+
+theorem exists_unique_hlift (hΓ₀ : ∀ y, f (Γ₀ y) = γ (0, y)) :
+    ∃! Γ : C(I × Y, E), ∀ y, Γ (0, y) = Γ₀ y ∧ f ∘ (Γ ⟨·, y⟩) = (γ ⟨·, y⟩) := by
```

Comment:
Why is this not stated as
```suggestion
    ∃! Γ : C(I × Y, E), ∀ y, Γ (0, y) = Γ₀ y ∧ f ∘ Γ = γ := by
```

Reply:
Good point. This means using a lot of `congr_fun` and the like but indeed it looks better.